### DISCUSSION TITLE
[vim9class] Super method dispatch; override concrete method with abstract method

### DISCUSSION THREAD

**Discussion by errael** [vim/vim#15463] [General]
The idea of this discussion is to sort out recent issues that have shown up regarding `vim9class`. Some may be spec, some bugs, some both, some may be related ... You are welcome to 
- comment
- give opinions on the the line items and how they should be dealt with
- suggest corrections about what issue (like #xxxx) is related to what line item
- suggest better description/wording about items
- note other **new** or related issues about the current vim9class implementation

This is about some loosely related items that expose issues with `vim9class`' method dispatch. See #15449, #15441, #15432, #15448; central to those issues is the idea of an abstract method overriding a concrete method; but there are also seemingly unrelated issues.

Items covered in this discuss (I will update this OP, and make note of changes at the end of this OP). The issues mentioned in this list are related to the line item, not necessarily 1-1.
1.  ~~Abstract method overrides concrete method.~~
     #15449
     #15432 

2.  ~~`super.method` may not execute in context of defining class.~~
     #15448
     #15477
3. ~~problem parsing abstract method declaration~~
    #15441 
    #15432
4. Signature of method in child class that overrides super class.
    #15442
    #15465
5. Dynamic method dispatch doesn't work if type is interface.
     #15484 
     #15495 
6. Missing return statement compiling abstract method declaration
     #15512
     There's a lost PR 15441 
7. problems with method dispatch: super -> abstract -> concrete
    #15514 

~~**(1) abstract method overrides concrete method**~~

Magically fixed; probably #15477

Something like the following is allowed in `Java` and I'm guessing most other compiled languages. While allowed in vim (it compiles) it causes problems.

```vim
vim9script

class A
    def _ToString(): string
        return 'A'
    enddef
endclass

abstract class B extends A
    abstract def _ToString(): string
endclass
```

Having a `this.` method and a `super.` method in an the object's method table is not all that unusual. While it's true that it's not handled correctly, having two method with the same name is not unusual.

Typically when a class defines an abstract method, there is no method with the same name in the class' superclass hierarchy. So the bugs probably come from not handling this previously untested situation.

1. the abstract method accessed with `this.`
2. the parent method accessed with `super.`

~~**(2) `super.method` may not execute in context of defining class**~~ PR #15477

But while exploring those  named issues I ran into this, nothing is abstract,
```vim
vim9script

class A
    def F(): string
        return 'AAAAA'
    enddef
endclass

class B extends A
    def F(): string
        return 'BBBBB'
    enddef
    def S(): string
        return super.F()
    enddef
endclass

class C extends B
    def F(): string
        return 'CCCCC'
    enddef
endclass

echo 'B.new().S():' B.new().S()         ### 'AAAAA'
echo 'C.new().S():' C.new().S()         ### 'BBBBB'
```
To be sure I wan't tricking myself, I tried this in `Java` and it works as expected. `AAAAA` is output in both cases. 

I think the underlying problem starts with the fact that when `C.new().S()` is encountered, the method `S()` is compiled and executed in `C`'s context. 

I had thought that the same compiled function from `B` would be used. If `B`'s compiled code references `A::F()` directly (without some dynamic dispatch), then that might work.

~~**(3) problem parsing abstract method declaration**~~ PR #15441

This is generally an unrelated item. But how it is fixed affects how some of the other issues manifest. What looks like the proper fix, makes it more difficult to diagnose some of the other problems mentioned in this discussion. So it is left in draft for now.

**(4) signature of method in child class that overrides super class** PR #15465

This is listed in this discussion because it was recently found, has to do with overriding methods, and there's discussion about spec/fix.

The original `vim9.1` spec says
> Object methods of the base class can be overruled.  The signature (arguments,
> argument types and return type) must be exactly the same.

But the spec does not match the implementation. To further complication things, the `vim-9.1` implementation causes data corruption.

**(5) Dynamic method dispatch doesn't work if type is interface**
Issue #15484. I wonder if this one can be safely fixed. (safe in terms of backwards compatibility)
```
vim9script

interface I
    def F(): string
endinterface

class A implements I
    def F(): string
        return 'A'
    enddef
endclass

class C extends A
    def F(): string
        return 'C'
    enddef
endclass

# Unexpected: following outputs 'A'. Expect 'C'
def TestI(i: I)
    echo i.F()
enddef
TestI(C.new())
```

---

**NOTES**
AFAICT, these are all issues in `vim-9.1` release.

@chrisbra @yegappan @dkearns @zzzyxwvut @LemonBoy, et al, comments?

PS This conceptually reminds me of #15112, but I've no idea if they are actually related.

- UPDATED Aug 10. Cleanup OP, list more seemingly related issues, clarify.
- UPDATED Aug 12. `super.Func()` fixed (AFAICT) by #15477
- UPDATED Aug 12. Note that for "abstract method override concrete method" having two methods with same name is normal.
- UPDATED Aug 12. Added #15484.
- UPDATED Aug 14. (2), (3) fixes merged
- UPDATED Aug 14 (1) fixed by recent PR, maybe #15477. 
   (5) has an experimental fix #15495

**Comment by errael** [comment]
Oops, should mention this is opened to make sure there are no spec issues. In particular, consider this
1. Adjust the spec and make the things that don't work compile time errors.
2. As time/effort permit, broaden the spec

I haven't studied @zzzyxwvut comments enough in some of the issues; but I believe there's mention that using/implementing `interface` solved some of the problems. I still don't have a handle on how many issues there are.

**Comment by errael** [comment]
Update and cleanup OP: list more seemingly related issues, clarifications, more descriptive. Comments welcome if you use `vim9class` and have information/opinion.

**Comment by errael** [comment]
About **(2) super.method may not execute in context of defining class**

The following shows some curious behavior; I don't know if/how it's related to any bugs. The way the disassembly changes seems suspicious. If anyone if familiar with the difference between `METHODCALL` and `DCALL`, why one gets used and not the other, ...

The following is start of all the examples in this particular post
```
vim9script

class A
    def String(): string
        return 'A'
    enddef
endclass

class B extends A
    def String(): string
        return super.String()
    enddef
endclass
```
**=== ex-1**
Disassembling `B.string()` gives
```
disa B.String
```
```
String
        return super.String()
   0 LOAD $0
   1 DCALL String(argc 0)
   2 RETURN
```
**=== ex-2**
But following the original with something that extends it, and then disassemble `B.String`, gives a different result
```
class C extends B
endclass

disa B.String
```
```
String
        return super.String()
   0 LOAD $0
   1 METHODCALL B.String(argc 0)
   2 RETURN
```
This seems very strange. Can this be right; a class method is compiled differently depending on if the class is extended? And notice that in the disassemble; it seems to say that `B.String` should be called, but in fact `A.String` is the target. Is this a bug in the disassembly?

**=== ex-3**
And disassembling before and after `C extends B` shows no change, which I presume means that once `B.String` is compiled it is not recompiled.
```
disa B.String

class C extends B
endclass

disa B.String
```
```
String
        return super.String()
   0 LOAD $0
   1 DCALL String(argc 0)
   2 RETURN
String
        return super.String()
   0 LOAD $0
   1 DCALL String(argc 0)
   2 RETURN
```

Fingers crossed that the disassembler is correct, as seen in output for ex-2, which would imply a compiler bug, and that fixing said hypothetical compiler bug actually fixes a problem. That's independent of getting different compiled code depending on the rest of the source code and when it's compiled.

**Comment by zzzyxwvut** [comment]
Regarding

> (3) problem parsing abstract method declaration

let me follow up my [initial](https://github.com/vim/vim/pull/15441#issuecomment-2276432849) findings:

> It feels kind of redundant [to repeat `implements`]
> since you do need to implement all abstract methods in order
> to have a concrete type so it is implied that some supertype
> has the `implements` clause. On the other hand, you do need
> to redeclare all `abstract` methods of an `interface` in an
> `abstract` class (or classes), so it is consistent and yet
> now it becomes optional for the clause (`E1091` was helpful
> before but, unfortunately, I failed to understand its source
> among 3 `abstract` `string()`s). Documentation entries under
> `E1349` and `E1389` and `:interface` do not tie in with the
> requirement for every concerned `class` to repeat `implements`
> for its hierarchy members.

about `interface`s and their  relation with the proposed patch  
([v1](https://github.com/vim/vim/commit/f958a72abd6df51789efdac23cfc5558847785d7), [v2](https://github.com/vim/vim/commit/4be245a767cd2460924c727cf7c343d30eba7c7e)).

I believe, the problem of whether `implements` needs to be  
repeated can be narrowed down to this set of conditions for  
either version of the patch:

1. An `abstract class` must declare an `abstract` method that is  
  inherited from an `interface` (e.g. `Parity` and `A` have `Odd()`  
  below).

2. There needs to be defined a script function that declares  
  its argument type to be of this kind of `interface` (e.g.  
  `Parity` in `Test*Parity()` below).  
  (I haven't yet  tested this setup with type methods.)

3. The `implements` clause must be repeated for a concrete  
  `class` that `extends` such an `abstract class` (e.g. `B` below).  
  Whether the `implements` clause must be repeated for every  
  other derived concrete `class` depends on, so it seems, the  
  implementation of a particular method that can be invoked  
  through such an `interface` type (e.g. `C` and `D`  implement  
  `Odd()` below).

If condition (3) is not met:

- With patch `v1`, you get an exception  
  `E1091: Function is not compiled:`.

- With patch `v2`, you get `0`.
  (which can be poked at with `type()` to emit some errors:
    ```
    E340: Internal error; if you can reproduce please report a bug
    E685: Internal error: f_type(UNKNOWN)
    ```
  ).

Because you are forced to repeat `implements` for some or  
every derived `class`, it seems to me that neither version of  
the patch is complete right now.

The following examples work with both `v1` and `v2`.  To break  
them, comment out `implements Parity` for `B` (`C`, `D`).

Then there is also a wrong-result success: comment out the  
stipulated by (3) `implements` clauses, BUT make `A` concrete,  
or keep `A` `abstract` and implement `Odd()`, then observe the  
uniform output of `TestOdd*()`.

#### (1) Plain `Parity` and friends.

```
vim9script

interface Parity
    def Odd(): bool
    def Even(): bool
endinterface

abstract class A implements Parity
    abstract def Odd(): bool

    def Even(): bool
        return !this.Odd()
    enddef
endclass

class B extends A implements Parity
    def Odd(): bool
        return false
    enddef
endclass

class C extends B implements Parity
    def Odd(): bool
        return true
    enddef
endclass

class D extends C implements Parity
    def Odd(): bool
        return false
    enddef
endclass

echo false == B.new().Odd()
echo true == B.new().Even()
echo true == C.new().Odd()
echo false == C.new().Even()
echo false == D.new().Odd()
echo true == D.new().Even()

####

def TestOddParity(p: Parity)
    echo p.Odd()
enddef

def TestEvenParity(p: Parity)
    echo p.Even()
enddef

def TestOddA(a: A)
    echo a.Odd()
enddef

def TestEvenA(a: A)
    echo a.Even()
enddef

def TestOddB(b: B)
    echo b.Odd()
enddef

def TestEvenB(b: B)
    echo b.Even()
enddef

def TestOddC(c: C)
    echo c.Odd()
enddef

def TestEvenC(c: C)
    echo c.Even()
enddef

def TestOddD(d: D)
    echo d.Odd()
enddef

def TestEvenD(d: D)
    echo d.Even()
enddef


echo 'TESTING TestEven*()'

TestEvenD(D.new())

TestEvenC(D.new())
TestEvenC(C.new())

TestEvenB(D.new())
TestEvenB(C.new())
TestEvenB(B.new())

TestEvenA(D.new())
TestEvenA(C.new())
TestEvenA(B.new())


TestEvenParity(D.new())
TestEvenParity(C.new())
TestEvenParity(B.new())

echo 'TESTING TestOdd*()'

TestOddD(D.new())

TestOddC(D.new())
TestOddC(C.new())

TestOddB(D.new())
TestOddB(C.new())
TestOddB(B.new())

TestOddA(D.new())
TestOddA(C.new())
TestOddA(B.new())


# Uncomment "implements Parity" for B, C, D!
TestOddParity(D.new())
TestOddParity(C.new())
TestOddParity(B.new())
```

#### (2) Traced `Parity` and friends.

```
vim9script

interface Parity
    def Odd(): bool
    def Even(): bool
endinterface

var traces: list<string> = []

abstract class A implements Parity
    abstract def Odd(): bool
    abstract def _Ddo(): bool

    def Even(): bool
        add(traces, this.string())
        return !this._Ddo()
    enddef

    def string(): string
        return typename(this)
    enddef
endclass

class B extends A implements Parity
    def Odd(): bool
        add(traces, this.string())
        return this._Ddo()
    enddef

    def _Ddo(): bool
        return false
    enddef
endclass

class C extends B
    def _Ddo(): bool
        return true
    enddef
endclass

class D extends C
    def _Ddo(): bool
        return false
    enddef
endclass

echo false == B.new().Odd()
echo true == B.new().Even()
echo true == C.new().Odd()
echo false == C.new().Even()
echo false == D.new().Odd()
echo true == D.new().Even()

add(traces, '::: 1 :::')

####

def TestOddParity(p: Parity)
    echo p.Odd()
enddef

def TestEvenParity(p: Parity)
    echo p.Even()
enddef

def TestOddA(a: A)
    echo a.Odd()
enddef

def TestEvenA(a: A)
    echo a.Even()
enddef

def TestOddB(b: B)
    echo b.Odd()
enddef

def TestEvenB(b: B)
    echo b.Even()
enddef

def TestOddC(c: C)
    echo c.Odd()
enddef

def TestEvenC(c: C)
    echo c.Even()
enddef

def TestOddD(d: D)
    echo d.Odd()
enddef

def TestEvenD(d: D)
    echo d.Even()
enddef


echo 'TESTING TestEven*()'

TestEvenD(D.new())

TestEvenC(D.new())
TestEvenC(C.new())

TestEvenB(D.new())
TestEvenB(C.new())
TestEvenB(B.new())

TestEvenA(D.new())
TestEvenA(C.new())
TestEvenA(B.new())

add(traces, '::: 2 :::')

TestEvenParity(D.new())
TestEvenParity(C.new())
TestEvenParity(B.new())

add(traces, '::: 3 :::')
echo 'TESTING TestOdd*()'

TestOddD(D.new())

TestOddC(D.new())
TestOddC(C.new())

TestOddB(D.new())
TestOddB(C.new())
TestOddB(B.new())

TestOddA(D.new())
TestOddA(C.new())
TestOddA(B.new())

add(traces, '::: 4 :::')

# Uncomment "implements Parity" for B!
TestOddParity(D.new())
TestOddParity(C.new())
TestOddParity(B.new())

echo traces
```

**Reply by errael** [reply]
> I believe, the problem of whether `implements` needs to be repeated can be narrowed down to this set of conditions for either version of the patch:
> 
>     1. An `abstract class` must declare an `abstract` method that is
>        inherited from an `interface` (e.g. `Parity` and `A` have `Odd()`
>        below).
> 
>     2. There needs to be defined a script function that declares
>        its argument type to be of this kind of `interface` (e.g.
>        `Parity` in `Test*Parity()` below).
>        (I haven't yet  tested this setup with type methods.)
> 
>     3. The `implements` clause must be repeated for a concrete
>        `class` that `extends` such an `abstract class` (e.g. `B` below).
>        Whether the `implements` clause must be repeated for every
>        other derived concrete `class` depends on, so it seems, the
>        implementation of a particular method that can be invoked
>        through such an `interface` type (e.g. `C` and `D`  implement
>        `Odd()` below).

Just to point out that the "add implements" thing has little to do with the
above conditions. Also nothing to do with the "missing return" issue.

This example has none of these features, it compiles without any patch,
including on vim-9.1. The following example is a simplification of #15484.

And adding `implements I` to `class C` gets it to dispatch correctly.

```
vim9script

interface I
    def F()
endinterface

class A implements I
    def F()
        echo 'A'
    enddef
endclass

class C extends A
    def F()
        echo 'C'
    enddef
endclass

# Unexpected: following outputs 'A'. Expect 'C'
def TestI(i: I)
    i.F()
enddef
TestI(C.new())
```

**Reply by zzzyxwvut** [reply]
> Just to point out that the "add implements" thing has little to do with the
> above conditions. Also nothing to do with the "missing return" issue.

I agree that the repetition of `implements` is a separate  
problem.  _At the time_ it was unclear whether the proposed  
`abstract` patches play any role in its exposure and _that  
list of conditions_ was made so that their shortcomings, real  
or imaginary, can be reproduced, and also it was done in  
response to your note [above](https://github.com/vim/vim/discussions/15463#discussioncomment-10290393):

> I haven't studied @zzzyxwvut comments enough in some of the issues; but I believe there's mention that using/implementing interface solved some of the problems.

so that the solved problems can be clearly identified.

**Comment by errael** [comment]
Added another issue to OP #15484.

(5) Dynamic method dispatch doesn't work if type is interface

**Comment by errael** [comment]
All the items have been addressed. Whew.

**Comment by errael** [comment]
This discussion is complete, AFAICT. I'm moving on but I'll leave this open for a little while since the item 
```
(5) Dynamic method dispatch doesn't work if type is interface
```
Needs to get sorted out. See the OP for issue/PR.

**Reply by zzzyxwvut** [reply]
It probably needs to be spelled out in some form in the OP  
(the discussion title is fine), somewhere below the line of  
"Magically fixed.", that the general question put forward is  
whether there ought to be support for `abstract class`es  
extending concrete `class`es, e.g.:

```
vim9script

class A
endclass

abstract class B extends A
endclass

class C extends B
endclass

C.new()
```

Because there is no requirement for `abstract class`es to  
declare `abstract` methods.  And if so, then `abstract`  
methods must be allowed to override concrete methods for  
consistency.  Otherwise, both `B` and `C` should be marked  
as invalid and rejected during compilation.

**Reply by errael** [reply]
> It probably needs to be spelled out in some form ...

I'm not sure what you mean. Is there a bug related to this?

**Reply by zzzyxwvut** [reply]
Having re-read myself section (1) in the OP, I understand  
your confusion at my suggestion and, in turn, am confused as  
well with _its_ contents.  If I'm not misremembering, the only  
published examples where concrete methods are overridden by  
`abstract` ones are these [two](https://github.com/vim/vim/issues/15449#issue-2454456919) and the (1) above.  And  [this  
comment](https://github.com/vim/vim/issues/15449#issuecomment-2274815542) questions whether the posted examples follow valid  
syntax.  Now with [your prompt aid](https://github.com/vim/vim/pull/15477), `super` behaves correctly.
So the (1) example can support this change:

```
vim9script

class A
    def ToString(): string
        return 'A'
    enddef
endclass

abstract class B extends A
#   abstract def ToString(): string
endclass

class C extends B
    def ToString(): string
        return super.ToString()
    enddef
endclass

echo C.new().ToString()
```

And as long as this is valid and supported syntax, the  
question about overriding validity of concrete methods by  
`abstract` ones should not be answered without taking into  
account the support of concrete `class`es being overridden by  
`abstract` ones for the sake of consistency.  But any attempt  
to invoke an `abstract` method through `super` is always  
invalid.

**Comment by errael** [comment]
When I try the above code I get the `missing return`. I hope that issue report is still open, I think the fix is closed without merge, but maybe some day a fix will be merged.

**Reply by zzzyxwvut** [reply]
With your patch (`27a302`) from #15495, which contains the  
reversal commit `v9.1.0675`, I see:

```
:class: created 'A'
:class: created 'B'
:class: created 'C'
A
```

Is your `abstract def ToString(): string` commented out in `B`?

**Reply by errael** [reply]
`A` looks correct, doesn't it?

**Reply by zzzyxwvut** [reply]
Yes, you should see `A`.

**Reply by errael** [reply]
So no bugs, right?

**Reply by zzzyxwvut** [reply]
No bugs.

**Comment by errael** [comment]
Added two more items to OP. They are explicit about things that are only hinted elsewhere. The idea is to show specific issues.

**Comment by errael** [comment]
@chrisbra

In the OP, (4), (6) have patches that were closed by some cleanup I did.

(7) is newly discovered and the issue has a `.patch.zip` which includes
the fix for (6).

(5) [vim9class] Unexpected method dispatch when type is interface #15484 
(InterfaceMethodDispatch)

(5) is a serious issue and has no fix. There are two solutions discussed;
one of them is clearly superior, but at first glance requires some
re-architecting (perhaps only tracking some additional information in class_T).
The other solution is simple, but requires additional "implements xxx" statements,
and there no guarantee that the interfaces are accessible; not backwards
compatible; better than nothing since it avoid erroneous behavior.
#15495, has some tracing (by `#ifdef`) and  some pointers related to possible fixes. 
#15484 hassome detailed discussion and the PR as patch file.

**Reply by chrisbra** [reply]
Thanks, I'll have a closer look tomorrow and come back to you

**Comment by chrisbra** [comment]
So let me go through this list:

> 1) Abstract method overrides concrete method.
>      - https://github.com/vim/vim/issues/15449
>      - https://github.com/vim/vim/issues/15432

So #15449 was fixed, not sure by which patch...
Issue #15432 is still opened and the initial patch for this fix turned out to be sub-optimal :( 

> 2.  ~~`super.method` may not execute in context of defining class.~~
>      #15448
>      #15477

#15448 and #15477 have both been fixed by commit: 58c957943030b4dc551bf210b76cb2882182260c so good.

> 3. ~~problem parsing abstract method declaration~~
>     #15441 
>     #15432

This seems to be a duplicate of 1) #15441 was the wrongly merged patch from last week  which should have fixed #15432 but since we reverted it, it is still open. 


> 4. Signature of method in child class that overrides super class.
>     #15442
>     #15465

We have some disagreements on #15442 whether this is a bug or not. Hm, now that I re-read it in detail, I tend to agree that the current behaviour is intentionally and thus this is not a bug and thus #15465 is also closed. 

> 
> 5. Dynamic method dispatch doesn't work if type is interface.
>      #15484 
>      #15495 

Agree, #15484 is serious and I wouldn't like the fix to add additional `implements` to workaround it (It really is just a work-around).

> 6. Missing return statement compiling abstract method declaration
>      #15512
>      There's a lost PR 15441 


hm, so should that patch #15512 be merged? I thaught this still had some issues?

> 7. problems with method dispatch: super -> abstract -> concrete
>     #15514 

There is a patch attached, but I am not sure what the status is. Should that be merged? Is that one good?

Did I forget anything?

**Reply by errael** [reply]
> So let me go through this list:
(1), (2), (3) are finished, AFAIK.
> 
> > 1. Abstract method overrides concrete method.
> >    
> >    * [[vim9class] Methods with a `super` call get lost in method dispatch #15449](https://github.com/vim/vim/issues/15449)
> >    * [[vim9class] Illegal lookup with upcasting for instance methods #15432](https://github.com/vim/vim/issues/15432)
> 
> So #15449 was fixed, not sure by which patch... Issue #15432 is still opened and the initial patch for this fix turned out to be sub-optimal :(

(1) - finished

Issue #15432 was a catch all report which included many things. One of the reasons this discussion was opened was to sort things out.

I'm not sure to which patch you are referring; probably the "missing return" patch, now a seperate item (6). This issue was closed when the original missing return patch was merged. It was reopened when the merge we reverted. BTW, the original "missing return" patch, #15441, is good and has no problems in and of itself (6) is supposed to be an identical copy; other things arose and/or were exposed when this was merged, summarized in https://github.com/vim/vim/pull/15441#issuecomment-2276670984, but the patch itself is OK.

> 
> > 2. ~`super.method` may not execute in context of defining class.~
> >    [[vim9class] Non-overridden methods with a `super` call recurse #15448](https://github.com/vim/vim/issues/15448)
> >    [Function dispatch problems when using super. #15477](https://github.com/vim/vim/pull/15477)
> 
> #15448 and #15477 have both been fixed by commit: [58c9579](https://github.com/vim/vim/commit/58c957943030b4dc551bf210b76cb2882182260c) so good.

(2) - finished

> 
> > 3. ~problem parsing abstract method declaration~
> >    [[vim9class] During compilation, don't look for a return statement for an `abstract` method. #15441](https://github.com/vim/vim/pull/15441)
> >    [[vim9class] Illegal lookup with upcasting for instance methods #15432](https://github.com/vim/vim/issues/15432)
> 
> This seems to be a duplicate of 1) #15441 was the wrongly merged patch from last week which should have fixed #15432 but since we reverted it, it is still open.

#15441 had too much stuff in it.

IMHO, there was no wrongly merged patch as discussed above under (1). That patch is now separated out into (6)

> 
> > 4. Signature of method in child class that overrides super class.
> >    [[vim9class] return type of method in child class not properly enforced #15442](https://github.com/vim/vim/issues/15442)
> >    [[vim9class] Only allow child class override method signatures that require no runtime checks. #15465](https://github.com/vim/vim/pull/15465)
> 
> We have some disagreements on #15442 whether this is a bug or not. Hm, now that I re-read it in detail, I tend to agree that the current behaviour is intentionally and thus this is not a bug and thus #15465 is also closed.

I strongly disagree, and for several reasons. So be it. 

(It's quite disappointing that none of the technical things I mentioned were ever addressed in the issue.)

> 
> > 5. Dynamic method dispatch doesn't work if type is interface.
> >    [[vim9class] Unexpected method dispatch when type is interface #15484](https://github.com/vim/vim/issues/15484)
> >    [[vim9class] DRAFT-DO-NOT-MERGE Incorportate interfaces from extended class #15495](https://github.com/vim/vim/pull/15495)
> 
> Agree, #15484 is serious and I wouldn't like the fix to add additional `implements` to workaround it (It really is just a work-around).

The incomplete patch (some decisions have to be made before it can be completed) provides some debug trace that might be useful.

> 
> > 6. Missing return statement compiling abstract method declaration
> >    [[vim9class] `E1027: Missing return statement` compiling abstract method declaration #15512](https://github.com/vim/vim/issues/15512)
> >    There's a lost PR 15441
> 
> hm, so should that patch #15512 be merged? I thaught this still had some issues?

As I've said, I believe the patch is OK; and that issues arose when this bug was fixed. Those issues are around (5) I think.

> 
> > 7. problems with method dispatch: super -> abstract -> concrete
> >    [[vim9class] problems with method dispatch: super -> abstract -> concrete #15514](https://github.com/vim/vim/issues/15514)
> 
> There is a patch attached, but I am not sure what the status is. Should that be merged? Is that one good?

I believe so.

> 
> Did I forget anything?

I just wanted to be sure that you are aware of these issues and the patches. Perhaps someone will come along interested in working on these issues and will find this stuff useful. I've got things cleaned up and unsubscribed; now for ignore. Be well.

**Reply by errael** [reply]
Forgot to mention, it goes without saying, but we all know what assume means. If there are bugs/issues around code I wrote then send email with issue number.

**Comment by zzzyxwvut** [comment]
There are two problems discussed in #15432 (1).  It is not  
yet clear for the first problem whether upcasting of types  
with `<>` (:help type-casting) [works as documented](https://github.com/vim/vim/issues/15432#issuecomment-2272298794), or does  
not work as documented, or is not documented at all.

**Reply by errael** [reply]
> There are two problems discussed in #15432 (1). It is not yet clear for the first problem whether upcasting of types with `<>` (:help type-casting) [works as documented](https://github.com/vim/vim/issues/15432#issuecomment-2272298794), or does not work as documented, or is not documented at all.

If there is a bug with `<{type}>` reporting it in a separate issue is probably best. Putting different things in the same issue easily leads to confusion.
